### ISSUE TITLE
The fn:build-uri function needs to perform URI encoding for path and query segments

### ISSUE THREAD

**Issue by ndw** [qt4cg/qtspecs#389]
The `fn:parse-uri` function describes decoding, but the `fn:build-uri` function fails to encode.

**Comment by ChristianGruen** [comment]
> The `fn:parse-uri` function describes decoding, but the `fn:build-uri` function fails to encode.

I re-read the original requirements of the URI functions in #72. There’s an example in the spec that describes the parsing of a file URI:

```
fn:parse-uri("file:/C:/Program%20Files/test.jar")`
```

It returns `C:/Program%20Files/test.jar` as path, which feels like we’ve stopped halfway through the journey. What would need to be done to get the decoded file path representation (`C:/Program Files/test.jar`)? Is it currently possible at all, or would a counterpart of `fn:encode-for-uri` still be helpful?

I remember the discussion, and it was mentioned that it will be a tough ride to get this defined. If we decide not to, we should possibly add some pseudocode to indicate how correctly encoded paths with percent characters can be rewritten to their decoded representation (e.g. with `fn:integer-to-string`, after #241 has been finalized).

The [EXPath File Module](http://expath.org/spec/file) is used a lot in practice to process file URIs. Most functions accept URIs starting with the file scheme, and [`file:path-to-uri`](http://expath.org/spec/file#fn.path-to-uri) can be used to encode a local file path to a URI. The module is – from today’s perspective – hopelessly under-specified, but maybe it helps to get a glimpse on possible use cases for `file:(/(/(/)))` URIs .

**Comment by ndw** [comment]
The current spec (with my PRs applied) says that `fn:parse-uri("file:/C:/Program%20Files/test.jar")` returns:

```
map {
  "uri": "file:/C:/Program%20Files/test.jar",
  "scheme": "file",
  "hierarchical": true(),
  "path": "/C:/Program%20Files/test.jar",
  "path-segments": array { "", "C:", "Program Files", "test.jar" }
}
```
The path you want is available from the `path-segments` array.

We could change the `path` so that it was `C:/Program Files/test.jar` but I was leaving the parts you couldn't blindly encode in their original form.

Or do I misunderstand?

**Comment by ChristianGruen** [comment]
Thanks, yes, I’ve seen that one. If I get it right, a Windows file path could be re-created as follows:

```xquery
let $uri := 'file:/C:/Program%20Files/test.jar'
return string-join(tail(parse-uri($uri)?path-segments, '/')
```

But I assume that `decode-uri($uri)` would be more intuitive?

**Comment by ChristianGruen** [comment]
Pardon for overloading this issue. I just noticed that in #390 the leading slash for Windows paths was questioned, so `array:tail` can be skipped.

Next, instead of `decode-uri`, it would rather be `decode-url` (instead of `decode-uri`) that could be standardized, as there is an equivalent function for that in MarkLogic, eXist-db and BaseX (as indicated in #72). I don’t know how this would currently be done in Saxon.

**Comment by ndw** [comment]
With respect to the leading "/" in the path part, `/C:/path`, I don't think it's wrong. We could certainly special case it away, but that's part of the URI and the leading slash is consistent with other file: URIs.

I wonder if we should add a `filename` property that returns the filesystem/operating system specific filename. That would be the same as the `string-join` example you show above but precomputed for the user. The problem with applying decoding to the path part is that you can't reverse it. `a/b%2Fc/d` becomes `a/b/c/d` and you can't reconstruct the path part from that if you need to interact with the resource as URI again.

**Comment by ndw** [comment]
The purist in me wants to keep the parts that are components of the URI in the form that's valid for the URI. In particular, note that `file://` + `/C:/path/` yields `file:///C:/path/` which is valid and reasonable where `file://C:/path` is not. And for the same reason, I'm reluctant to change `/path/a%2Fb/parts` to `/path/a/b/parts` because it's impossible to reconstruct the original and if this was an HTTP URI, for example, you need the original form if you want to intract with the web. `/path/a%2Fb/parts` and `/path/a/b/parts` are completely different URIs.

**Comment by ChristianGruen** [comment]
Closed (#566 is a follow-up issue).
